### ISSUE TITLE
Allow config variable to be set by user on Connect

### ISSUE THREAD

**Issue by gadenbuie** [jjallaire/config#46]
Since `R_CONFIG_ACTIVE` is set without user control to `rsconnect` *and* Connect disallows app-specific overrides of this variables, it is impossible to deploy the same app to Connect with different parameters.

Suppose, for example, I want to deploy a development version and a production version of the same app to Connect. This would let me do some app testing before pushing big changes to production users.

But because I don't have control over the value of `R_CONFIG_ACTIVE` and because Connect won't let me set that environment variable, there's no way to differentiate these two configurations.

I think it makes sense for `R_CONFIG_ACTIVE` to have the default value of `rsconnect`. Ideally, Connect would let users make changes to that environment variable. But it would probably be easier to add second env var that can be checked first, e.g. `R_CONFIG_ACTIVE_APP`.

**Comment by gadenbuie** [comment]
Also, you mentioned in #43 that this is documented, but it's not advertised in the README. Searching config's pkgdown site for "rsconnect" also doesn't return any hits. I think this should be more prominently documented that the config described in the README won't work as described on Connect.

**Comment by andrie** [comment]
I'll respond in more detail later, but I've pushed my dev branch into main, so now the Connect vignette is live at https://rstudio.github.io/config/articles/rsconnect.html

**Comment by gadenbuie** [comment]
Thanks @andrie. The appeal of using config, for me, is being able to globally change a set of constants by changing the value of an envvar or by updating a config file. It's helpful that `config::get()` lets me provide a `config` with each call, but I'd much rather that be done globally for all calls. Here's what I ended up including in my app:

```r
config_use <- function(config = "default", .local_envir = parent.frame()) {
  withr::local_envvar(
    list(R_CONFIG_ACTIVE_APP = config),
    .local_envir = .local_envir
  )
}

config_active <- function(default = "default") {
  Sys.getenv(
    "R_CONFIG_ACTIVE_APP",
    Sys.getenv("R_CONFIG_ACTIVE", default)
  )
}

config_is_active <- function(config) {
  identical(config, config_active())
}

config_get <- function(value, config = config_active(), ...) {
  config::get(value, config = config, ...)
}
```

I guess the above comes with two feature requests:

1. `config::active()`: A function that returns the active configuration, rather than having to directly query if a certain config name is active with `config::is_active()`.
2. `config::use()`: A function that sets the current configuration locally. I used {withr} which would scope the change to the current function call or the current R session.

**Comment by andrie** [comment]
I probably misunderstand what you're trying to do here.  But in essence it seems you want to have a single `config.yml` and a single env var that controls all variables on Connect.

I believe this is possibly using the existing package:
 * Use inheritance in your `config.yml`
 * Point `config::get()` to a different env var, using the `config` argument of `config.get()`

I have updated the vignette to include a mention of inheritance (https://rstudio.github.io/config/articles/rsconnect.html)

PS. I took another look at the `config::is_active()` and it seems to me this function needs a `config` argument to conform to the usage of `config::get()` that I just described.

**Comment by gadenbuie** [comment]
Thank @andrie. Let me try again. I definitely want to use a single `config.yml` file.

I opened this issue because the `config` argument of `config::get()` reads `R_CONFIG_ACTIVE` by default and Connect doesn't allow users to change this value. (Connect admins _are allowed_ to change the value globally, but standard practice is for this value to be `"rsconnect"`.) This means that only one active configuration is allowed on Connect<sup>1</sup>.

The work-around is to look up the active configuration using another envvar that isn't limited on Connect, e.g. `R_CONFIG_ACTIVE_APP`. To make this easier for myself, I created local proxies for key {config} functions, but this could all be rolled into {config}. Here, I added `config_active()`:

```r
config_active <- function(default = "default") {
  Sys.getenv(
    "R_CONFIG_ACTIVE_APP",
    Sys.getenv("R_CONFIG_ACTIVE", default)
  )
}
```

This could be brought into {config} as `config::active()` or `config::active_configuration()`. It uses the same envvar pattern as used by {config} currently, but first looks in `R_CONFIG_ACTIVE_APP` before consulting `R_CONFIG_ACTIVE`.

At this point, I could now use

```r
config::get(value, config = config_active())
```

everywhere in my code. But {config} could also update the default value of the `config` argument to use `config::active()` or `config::active_configuration()`. I'd rather do this than to have to include my custom `config` argument in every call to `config::get()`, so I wrote this wrapper:

```r
config_get <- function(value, config = config_active(), ...) {
  config::get(value, config = config, ...)
}

# now I can use config_get() just like config::get()
config_get("db")
```

That's the core idea. But then I probably muddied the waters by including  two feature requests and some extra code:

1. As a user, sometimes I'd like to know the name of the active configuration. Internally to {config} it'd also be helpful to have one function that implements how the active configuration is determined.

2. As a user, I'd like to be able to set, possibly temporarily, the active configuration. `config_use()` in my last comment demos a simple way to achieve this with {withr}. In {config}, that function would likely want to take `config` and `file` and provide an easy way to temporarily choose which value is used (by default) across all `config::get()` calls.

3. Finally, my `config_is_active()` shows how `config::is_active()` could be updated to use function described in point (1).

***

1. Why is setting `R_CONFIG_ACTIVE` disallowed for individual apps on Connect? If this restriction were lifted, most of the discussion in this issue would go away...

**Comment by andrie** [comment]
It seems to me easier to simply have a wrapper around `config::get()`, like this:

``` r
cfg_get_app <- function (
  value = NULL,
  config = Sys.getenv("R_CONFIG_ACTIVE_APP", "default"),
  file = Sys.getenv("R_CONFIG_FILE", "config.yml"),
  use_parent = TRUE
) {
  config::get(value = value, config = config, file = file, use_parent = use_parent)
}
```

**Comment by andrie** [comment]
PS.  I'm pondering what you did.  Your suggestion may actually just contain a critical insight that could mean we can make this change backward compatible.

**Comment by gadenbuie** [comment]
> It seems to me easier to simply have a wrapper around `config::get()`

I agree with you here, actually. Really, I think Connect shouldn't block app-level user changes to `R_CONFIG_ACTIVE`, but that behavior's been around for ~5 years, {config} might also need to help rectify the situation.

Fundamentally, though, my `config_get()` wrapper is the same as your `cfg_get_app()`. Personally, I'd recommend authors wrapping an external package take on as little of the wrapped package's API, so the key difference is I used `...` to pass options I didn't want to modify down to `config::get()`.

**Comment by gadenbuie** [comment]
> PS. I'm pondering what you did. Your suggestion may actually just contain a critical insight that could mean we can make this change backward compatible.

Thanks! I'd be happy to prepare a PR or two if you'd like.

**Comment by andrie** [comment]
In a conversation with the Posit Connect development team, it transpires that an imminent update to Connect will allow app owners to modify the `R_CONFIG_ACTIVE` env var at the app level.

This means that the workarounds described in this issue are only relevant to users of Connect that have not yet updated to a version of Connect to be release around April 2023.

For this reason we will leave the workaround as described in the vignette and this issue, but won't change `config` itself to make this any easier.
